### ISSUE TITLE
Regression — doesn't use the bundled Yarn version

### ISSUE THREAD

**Issue by koistya** [GoogleCloudPlatform/buildpacks#175]
The recently released buildpack for Node.js broke the way `yarn install` works — it's no longer uses the bundled yarn version (defined in `.yarnrc.yml` file). At least, providing a patched Yarn CLI as a workaround for #174 doesn't work.

Project sample — https://github.com/kriasoft/relay-starter-kit → `api` (GraphQL API for GCF environment)

This Cloud Function is being bundled with Yarn v3.2.0 CLI and NPM packages (only production dependencies) as a deploy time optimization. After you run `yarn api:build`, the build output looks like this:

```bash
/api/dist
 ├── .yarn/            # Yarn CLI and bundled NPM packages (dependencies)
 ├── .pnp.cjs          # Yarn's PnP module resolution
 ├── .pnp.loader.mjs   # ESM loader with PnP support
 ├── .yarnrc.yml       # Yarn configuration (version number, CLI path `.yarn/releases/yarn-{version}.cjs`)
 ├── index.js          # The bundled application code (via Rollup)
 ├── package.json      # Includes Node.js package manager info — "packageManager": "yarn@3.x.y"
 └── yarn.lock         # The list of production ONLY dependencies
                       # So, there is no need to run `yarn workspaces focus` during GCF bootstrapping
```

Then when your run `yarn api:deploy` it deploys the build output from `./api/dist` folder to GCF environment.

Note that `yarn.lock`, `.pnp.cjs` files from GCF source bundle might not be compatible with the Yarn version lately being enforced by the buildpack. Also, it's currently not possible to implement workarounds for other regressions such as #174.

**Buildpack Version**: `base_20220114_18_04_RC00`
**Ref**: https://github.com/GoogleCloudPlatform/buildpacks/commit/083551c792d7de30f40242d07b93164507a840dd

Also, [caching](https://github.com/kriasoft/relay-starter-kit/blob/71462f8c707200b81800c13063147e01799dc6e6/.github/workflows/pull_request.yaml#L24-L31) still doesn't work correctly with Yarn-based GCFs (`DEBUG: ***** CACHE MISS: "prod dependencies"`).  And, deployment speed is expected to be much better when deploying GCF code already bundled with all the prod dependencies.

**Comment by matthewrobertson** [comment]
In theory, https://github.com/GoogleCloudPlatform/buildpacks/commit/083551c792d7de30f40242d07b93164507a840dd should have added support for Yarn2 / PnP mode, so I am surprised that this was working before.

The error message of `FAILURE and message: Unknown Syntax Error: Invalid option name ("--production=false").` sounds like the buildpack is failing to recognize that the project is using Yarn2. Otherwise it would be running `yarn install --immutable --immutable-cache`. The logic for this is here:

https://github.com/GoogleCloudPlatform/buildpacks/blob/647850eeea968f2a3926903a0e9466bf42d57428/pkg/nodejs/yarn.go#L39-L47

**Comment by matthewrobertson** [comment]
Ahh so it seems like the commit to add Yarn2 support didn't go out in the most recent gcf release but https://github.com/GoogleCloudPlatform/buildpacks/commit/0c08a9c7992b50c941f8be8f49f74ffe12317a24 did.

This app builds fine using the code that is checked in to main. The reason that this appeared to work before is because we treated it like Yarn1 and just installed installed everything in `node_modules`. The new `--production` flag isn't compatible with Yarn2 so that is causing the regression, which will be fixed in the next release that will actually include Yarn2 support.

**Comment by matthewrobertson** [comment]
That said I am not sure apps using Yarn2 PnP mode will run on GCF because of the way we install and invoke the functions framework in production. Addressing that is still a WIP. 

To be clear, GCF has never (and still does not) support Yarn2 or Yarn2 PnP mode. I am actively working on it, but I would advise against using it for now.

**Comment by koistya** [comment]
Hm...  I have a bunch of projects for various clients using Yarn PnP that are currently broken, unable to update / re-deploy. Was using Yarn PnP just fine, by passing `NODE_OPTIONS=--require=./.pnp.cjs`  env var during deploy, it was installing dependencies using the bundled Yarn version into `.yarn` folder, and `./.pnp.cjs` was able to pick them up just fine.

https://github.com/kriasoft/relay-starter-kit/blob/71462f8c707200b81800c13063147e01799dc6e6/scripts/api-deploy.js#L49

**Comment by matthewrobertson** [comment]
@koistya https://github.com/GoogleCloudPlatform/buildpacks/commit/c5bf000ef66c97ac30671d465b4711d1649115e7 adds proper support for Yarn2 on GCF. It will go out in next week's release after which you should be able use Yarn2 without any modifications to `NODE_OPTIONS`. Are all your apps using the `nodejs16` runtime? I can probably expedite that release to unblock you.

**Comment by tilgovi** [comment]
I just ran into this today. I thought it was due to upgrading to nodejs16, but that may just be a coincidence. We have been deploying GCFs that we develop with Yarn 2 without issue. Today, we started hitting this new conditional path and incorrectly getting the Yarn 1 command line flags that cause our build to fail.

**Comment by matthewrobertson** [comment]
@tilgovi are you using Yarn2's PnP mode? I am expediting a release that should have the necessary changes to add support for Yarn2 to GCF. It will take a few hours to hit all GCP regions.

**Comment by tilgovi** [comment]
Yes, we're using PnP. It looks like the rollout fixed our issues. Thanks for the speedy response!

**Comment by matthewrobertson** [comment]
Awesome. Yarn2 PnP should be fully supported by all products that use our buildpacks. I am closing this issue. If you experience any new problems please file bugs.

**Comment by koistya** [comment]
@matthewrobertson there are a few more errors (unable to deploy a Node.js + Yarn project).

Q: Is `NODE_ENV` variable is no longer being set to `production`? Previously, it was set to production and this (local dev) code block below was ignored in the GCF environment:

https://github.com/kriasoft/relay-starter-kit/blob/71462f8c707200b81800c13063147e01799dc6e6/api/index.ts#L46-L61

After the recent `buildpack` release, it fails to run the code using `process.env.NODE_ENV === "production"` check for detecting if the code is running on a production server vs local development machine.

**Comment by matthewrobertson** [comment]
@koistya can you share more details about how the project is configured? Is express listed in your `package.json` / `yarn.lock` and `.yarn/cache`?

**Comment by TimFHayes** [comment]
Hey Matthew,

I am having similar trouble deploying GCF written in Python. The log message is:

"Step #1 - "build": Unable to delete previous cache image: DELETE https://eu.gcr.io/v2/quinjo-image-handler/gcf/europe-west2/046fef8a-943d-4ca7-9c7e-ac5f2b3008a3/cache/manifests/sha256:f55b68d62ae8465c55517828f2b7fd8ebc0f943e4ff0c676024b828ca49bcaa0: GOOGLE_MANIFEST_DANGLING_TAG: Manifest is still referenced by tag: latest"
timestamp: "2022-03-05T20:13:00.505488881Z"

My deploy command from my local machine is:

gcloud functions deploy gcs_textualize_pdf --region europe-west2 --runtime python39 --trigger-resource quinjo_bucket_a --trigger-event google.storage.object.finalize

This has been working fine for two years. I cannot see a workaround - could you please offer some advice?

Many thanks.

Tim Hayes

**Comment by matthewrobertson** [comment]
@TimFHayes can you please open a new issue for this if you are continuing to encounter issues. It sounds like your GCR repo is in a bad state so you might try to manually delete `https://eu.gcr.io/v2/quinjo-image-handler/gcf/europe-west2/046fef8a-943d-4ca7-9c7e-ac5f2b3008a3/cache/manifests:latest` but hard to say without additional information

**Comment by TimFHayes** [comment]
Big thanks for replying Matthew.

The issue has been resolved but it is totally unconnected with the issue you posted on.

I have been deploying a Python function from a Windows PC. I utilised a requirements.txt file alongside main.py.

In migrating from another PC, requirements.txt was picked up from a different folder and lacked an entry for an additional import, namely dateparser. 

The gcloud error may be accurate, but it is not informative as to the problem. I only found it because I had a hunch, so to speak.

Tim
